websocket.go

```diff
@@ -21,6 +22,72 @@ import (
 	graceful "gopkg.in/tylerb/graceful.v1"
 )
 
+const certificateReloaderLeeway = 1 * time.Hour
+
+// CertificateReloader takes care of reloading a TLS certificate when
+// requested.
+type CertificateReloader struct {
+	sync.RWMutex
+	cert     *tls.Certificate
+	certPath string
+	keyPath  string
+}
+
+// NewCertificateReloader takes two file paths as parameter that contain
+// the certificate and the key data to create an automatic reloader. It will
+// try to read again the files when the certificate is almost expired.
+func NewCertificateReloader(certPath, keyPath string) (*CertificateReloader, error) {
+	loader := &CertificateReloader{
+		certPath: certPath,
+		keyPath:  keyPath,
+	}
+
+	return loader, loader.reload()
+}
+
+func (cr *CertificateReloader) reload() error {
+	newCert, err := tls.LoadX509KeyPair(cr.certPath, cr.keyPath)
+	if err != nil {
+		return err
+	}
+
+	cr.Lock()
+	cr.cert = &newCert
+	// Successful parse means at least one certificate.
+	cr.cert.Leaf, err = x509.ParseCertificate(newCert.Certificate[0])
+	cr.Unlock()
+
+	return err
+}
+
+// GetCertificateFunc makes a function that can be passed to the TLSConfig
+// so that it resolves the most up-to-date one.
+func (cr *CertificateReloader) GetCertificateFunc() func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	return func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		cr.RLock()
+
+		exp := time.Now().Add(certificateReloaderLeeway)
+
+		// Here we know the leaf has been parsed successfully as an error
+		// would have been thrown otherwise.
+		if cr.cert == nil || exp.After(cr.cert.Leaf.NotAfter) {
+			// Certificate has expired so we try to load the new one.
+
+			// Free the read lock to be able to reload.
+			cr.RUnlock()
+			err := cr.reload()
+			if err != nil {
+				return nil, err
```

Comment:
This return, plus the pending defer, will trigger a duplicate call to RUnlock(), which might be ok, I don't know.

I think mixing "defer unlock" and "need to just unlock this before this call and relock it after" is asking for trouble, I'm still looking and thinking if there's another way.

Reply:
The defer is declared after the if clause. It could actually be replaced by a copy of cert then unlock then return but it won't have a duplicate unlock.